src/Std/Internal/Async/Basic.lean

```diff
@@ -13,21 +13,241 @@ namespace Internal
+
+- **Monadic Interfaces**: These types provide a good way to to chain and manipulate context. These
+  can contain a `Task`, enabling manipulation of both asynchronous and synchronous code.
+- **Concrete Task Types**: Concrete units of work that can be executed within these contexts.
+
+## Monadic Interfaces
+
+These types provide a good way to to chain and manipulate context. These can contain a `Task`,
+enabling manipulation of both asynchronous and synchronous code.
+
+- `BaseAsync`: A monadic interface for infallible asynchronous computations (alias for
+  `EAsync Empty`)
+- `EAsync`: A monadic interface for asynchronous computations that may fail with an error of type
+  `ε`
+- `Async`: A monadic interface for IO-based asynchronous computations that may fail with `IO.Error`
+  (alias for `EAsync IO.Error`)
+
+## Concurrent Units of Work
+
+These are the concrete computational units that exist within the monadic contexts. These types
+should not be created directly.
+
+- `Task`: A task that is pure, so it returns a
+- `ETask`: A task that may fail with an error of type `ε`
+- `AsyncTask`: A task that may fail with an `IO.Error` (alias for `ETask IO.Error`)
+
+## Relation
+
+Concrete units of work that can be executed within these contexts.
+
+These types are related by two functions in the type classes `MonadAsync` and `MonadAwait`: `async`
+and `await`. The `async` function extracts a concrete asynchronous task from a computation inside
+the monadic context. In effect, it makes the computation run in the background and returns a
+task handle that can be awaited later. On the other side, the `await` function takes a task and
+re-inserts it into the monadic context, allowing the result to be composed using the monadic bind.
+This relation between `async` and `await` enables precise control over when a computation starts and
+when its result is actually used. You can spawn multiple asynchronous tasks using `async`, continue
+with other operations, and later rejoin the computation flow by awaiting their results.
+
+These functions should not be used directly. Instead, prefer higher-level combinators such as
+`race`, `raceAll`, `concurrently`, `background` and `concurrentlyAll`. The best way to think about
+how to write your async code, it to avoid using the concurrent units of work, and only use it when integrating
+with non async code that uses them.
+
+-/
+
+/--
+Typeclass for monads that can "await" a computation of type `t α` in a monad `m` until the result is
+available.
+-/
+class MonadAwait (t : Type → Type) (m : Type → Type) extends Monad m where
+  /--
+  Awaits the result of `t α` and returns it inside the `m` monad.
+  -/
+  await : t α → m α
+
+/--
+Represents monads that can launch computations asynchronously of type `t` in a monad `m`.
+-/
+class MonadAsync (t : Type → Type) (m : Type → Type) extends Monad m where
+  /--
+  Starts an asynchronous computation in another monad.
+  -/
+  async : m α → m (t α)
```

Comment:
Oh, something I forgot in my review. The spawning part of this API (e.g., `async`) does not provide any way to control the priorities of these tasks. Being able to mark a spawned `Task` as `Task.Priority.dedicated` can be quite important. I would recommend having an optional `prio` argument on `async` that is passed through the layers to the task spawning:

```suggestion
  async (x : m α) (prio := Task.Priority.default) : m (t α)
```